udf_js/lib/wtdstats.js

```diff
@@ -0,0 +1,75 @@
+// https://github.com/mozilla/glam/blob/e6d345d0e61df23549ece416e2d41e800980de41/src/utils/stats.js
+(function() {
+function sum(vs) { return vs.reduce((a, b) => a + b, 0); }
+
+ function ascending(a, b) {
+  if (a < b) return -1;
+  if (a > b) return 1;
+  if (a >= b) return 0;
+  return NaN;
+}
+
+ function byKey(k) {
+  return function sortByKey(a, b) {
+    if (a[k] < b[k]) return -1;
+    if (a[k] > b[k]) return 1;
+    if (a[k] >= b[k]) return 0;
+    return NaN;
+  };
+}
+
+ function orderedNumbers(values, valueof) {
+  return Float64Array.from(values.map(valueof)).sort(ascending);
+}
+
+ function cumSum(values, valueof = (v) => v) {
+  return values.map(valueof).reduce((acc, v, i) => {
+    if (i === 0) {
+      acc.push(v);
+    } else {
+      acc.push(acc[i - 1] + v);
+    }
+    return acc;
+  }, []);
+}
+
+ function nnInterp(x, y, xOut) {
+  // returns yOut by mapping x -> xOut, finding the index, then applying to y
+  // constant (nearest-neighbors) interpolation with f=1 (always right-sided)
+  // find ind in x where xOut values should below, then look at y[ind]
+  // assume x & y are pre-sorted and match.
+  const xInds = xOut.map((xi) => x.findIndex((xx) => xx > xi));
+  const low = Math.min(...x);
+  const high = Math.max(...x);
+  return xInds.map((i) => {
+    // left & right censor extreme values.
+    if (i === -1 && xOut[i] <= low) return y[0];
```

Comment:
I know this is just directly from glam but this function looks a little sketchy and is hard to understand.  I think the `xInds.map` logic is wrong.  If `i === -1`, then `xOut[-1]` is undefined and so these clauses are always false and y[i] is always returned which means y[-1] can be returned.  This is what's causing the `NaN` when weights are 1.  I think there's some other logic elsewhere that's causing the difference between weights 1 and 3 but I'm not sure where.  Probably something to talk to Hamilton about and maybe it works fine in the cases we need to use it in but I have trouble trusting this code.

Reply:
It might be worthwhile to wait on this PR then until more testing is done on the underlying code.

Reply:
That would be ideal as long as this doesn't block anything that's very time-sensitive.

Reply:
yes, I wouldn't probably move forward with this PR until we actually write [tests of my port](https://github.com/mozilla/glam/issues/351) & more thoroughly review the code. Moving away from the old `histogram.js` implementation is going to be a pretty big shift, so it is worth doing it right this time around.